inline-java/src/Language/Java.hs

```diff
@@ -4,38 +4,196 @@
+  unsafeUncoerce _ = error "unsafeUncoerce: value doesn't match target type."
+instance Coercible () 'Void where
+  coerce = error "Void value undefined."
+  unsafeUncoerce _ = ()
+
+-- | NULL terminate byte strings, because those that were not created from
+-- statically allocated literals aren't guaranteed to be.
+nullTerminate :: ByteString -> ByteString
+nullTerminate = (`BS.snoc` '\0')
+
+-- | FindClass() special cases class names: in that case it doesn't want
+-- a full signature, just the class name.
+signatureStrip :: ByteString -> ByteString
+signatureStrip sig | Just ('L', cls) <- BS.uncons sig = BS.init cls
+signatureStrip sig = sig
+
+-- | Creates a new instance of the class whose name is resolved from the return
+-- type.
+new
+  :: forall a sym.
+     ( Coerce.Coercible a (J ('Class sym))
+     , Coercible a ('Class sym)
+     , KnownSymbol sym
+     )
+  => [JValue]
+  -> IO a
+new args = do
+    let argsings = map jtypeOf args
+        voidsing = sing :: Sing 'Void
+    klass <- findClass (nullTerminate (signatureStrip (signature (sing :: Sing ('Class sym)))))
+    Coerce.coerce <$> newObject klass (nullTerminate (methodSignature argsings voidsing)) args
+
+-- | The Swiss Army knife for calling Java methods. Give it an object or
+-- any data type coercible to one, the name of a method, and a list of
+-- arguments. Based on the type indexes of each argument, and based on the
+-- return type, 'call' will invoke the named method using of the @call*Method@
+-- family of functions in the JNI API.
+--
+-- When the method name is overloaded, use 'upcast' or 'unsafeCast'
+-- appropriately on the class instance and/or on the arguments to invoke the
+-- right method.
+call
+  :: forall a b ty1 ty2. (Coercible a ty1, Coercible b ty2, Coerce.Coercible a (J ty1))
+  => a
+  -> ByteString
+  -> [JValue]
+  -> IO b
+call obj mname args = do
+    let argsings = map jtypeOf args
+        retsing = sing :: Sing ty2
+    klass <- findClass (nullTerminate (signatureStrip (signature (sing :: Sing ty1))))
+    method <- getMethodID klass mname (nullTerminate (methodSignature argsings retsing))
+    case retsing of
+      SPrim "boolean" -> unsafeUncoerce . coerce <$> callBooleanMethod obj method args
+      SPrim "byte" -> unsafeUncoerce . coerce <$> callByteMethod obj method args
+      SPrim "char" -> error "call: unimplemented"
+      SPrim "short" -> error "call: unimplemented"
+      SPrim "int" -> unsafeUncoerce . coerce <$> callIntMethod obj method args
+      SPrim "long" -> unsafeUncoerce . coerce <$> callLongMethod obj method args
+      SPrim "float" -> error "call: unimplemented"
+      SPrim "double" -> unsafeUncoerce . coerce <$> callDoubleMethod obj method args
+      SVoid -> do
+        callVoidMethod obj method args
+        -- Anything uncoerces to the void type.
+        return (unsafeUncoerce undefined)
+      _ -> unsafeUncoerce . coerce <$> callObjectMethod obj method args
+
+-- | Same as 'call', but for static methods.
```

Comment:
Right, let's focus on the type sig first. I'd be in favor of having:

``` haskell
callStatic :: forall a ty. Coercible a ty => String {- or ByteString -} -> ByteString -> [JValue] -> IO a
```

My main argument here is that users never have a `Sing sym` around already, so they have to make one up out of thin air anyway, so why bother with `sing :: Sing "my.java.Class"` when we could just write `"my.java.Class"`.

(I mentionned `reifySymbol` only because it would allow us to have the class name reified at the type level if we need it there)